src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/Program.cs

```diff
@@ -60,9 +63,55 @@ public static int Main(string[] args)
                 }
             }
 
+            return writeGrammar
+                ? WriteGrammarFile(inputFile, outputFile)
+                : WriteCSharpSourceFiles(inputFile, writeSource, writeTests, writeSignatures, outputFile);
+        }
+
+        private static int WriteUsage()
+        {
+            Console.WriteLine("Invalid usage:");
+            var programName = "  " + typeof(Program).GetTypeInfo().Assembly.ManifestModule.Name;
+            Console.WriteLine(programName + " input-file output-file [/test | /grammar]");
+            Console.WriteLine(programName + " input-file /sig");
+            return 1;
+        }
+
+        private static Tree ReadTree(string inputFile)
+        {
             var reader = XmlReader.Create(inputFile, new XmlReaderSettings { DtdProcessing = DtdProcessing.Prohibit });
             var serializer = new XmlSerializer(typeof(Tree));
-            Tree tree = (Tree)serializer.Deserialize(reader);
+            return (Tree)serializer.Deserialize(reader);
+        }
+
+        private static int WriteGrammarFile(string inputFile, string outputLocation)
+        {
+            try
+            {
+                var generator = new GrammarGenerator(ReadTree(inputFile));
+                var grammarText = generator.Run();
+
+                var outputMainFile = Path.Combine(outputLocation.Trim('"'), $"CSharp.generated.g4");
+
+                using var outFile = new StreamWriter(File.Open(outputMainFile, FileMode.Create), Encoding.UTF8);
+                outFile.Write(grammarText);
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine("Generating grammar failed.");
+                Console.WriteLine(ex);
+
+                // purposefully fall out here and don't return an error code.  We don't want to fail
```

Comment:
Why not treat this like other generated files (ie. fail the build)?

Reply:
I was a little nervous about that and didn't want to potentially cause a bunch of a pain the next time something changes in syntax.xml that for some reason i didn't consider.  Since this is an optional sort of artifact, it felt better to not just fail there.  That way compiler devs weren't blocked while trying to actually make important engineering changes.  And we could then asynchronously figure out how to update the grammar generator.

That was my thought process at least.  :)

Reply:
OK. We can adjust as we go.